File: October/11/distinctsubseq.py
# LeetCode 115: Distinct Subsequences
# Explanation:
# 1. Count how many distinct subsequences of `s` equal string `t`.
# 2. Use dynamic programming:
#    - dp[i][j] = number of ways first i chars of `s` form first j chars of `t`.
#    - If s[i-1] == t[j-1], then:
#         dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
#      Else:
#         dp[i][j] = dp[i-1][j]
# 3. Initialize dp[i][0] = 1 since empty t is always a subsequence.
# Time Complexity: O(m * n)
# Space Complexity: O(m * n)

class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m, n = len(s), len(t)
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        
        for i in range(m + 1):
            dp[i][0] = 1
        
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if s[i - 1] == t[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j]
                else:
                    dp[i][j] = dp[i - 1][j]
        
        return dp[m][n]

File: 2025/October/11/distinctsubseq.java
// LeetCode 115: Distinct Subsequences
// Explanation:
// 1. dp[i][j] = number of ways s[0..i-1] can form t[0..j-1].
// 2. Base: dp[i][0] = 1, since empty t is subsequence of any prefix of s.
// 3. Transition:
//    - If s[i-1] == t[j-1], dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
//    - Else dp[i][j] = dp[i-1][j]
// Time Complexity: O(m * n)
// Space Complexity: O(m * n)

class Solution {
    public int numDistinct(String s, String t) {
        int m = s.length(), n = t.length();
        long[][] dp = new long[m + 1][n + 1];
        
        for (int i = 0; i <= m; i++)
            dp[i][0] = 1;
        
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s.charAt(i - 1) == t.charAt(j - 1))
                    dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
                else
                    dp[i][j] = dp[i - 1][j];
            }
        }
        return (int) dp[m][n];
    }
}

// System.out.println(new Solution().numDistinct("rabbbit", "rabbit"));
// Output: 3

// System.out.println(new Solution().numDistinct("babgbag", "bag"));
// Output: 5

